### PR TITLE
Add `EnablePrivateLinkFastPath` to the Express Route Connection resource

### PR DESCRIPTION
**Pull request by fjaeckel**
## Community Note

* Please vote on this PR by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original PR to help the community and maintainers prioritize for review
* Please do not leave "+1" or "me too" comments, they generate extra noise for PR followers and do not help prioritize for review

## Description

Add a new optional `EnablePrivateLinkFastPath` option to the Express Route Connection resource. This enables users to enable the feature via Terraform on their Express Route connection. 

I chose `enable_private_link_fast_path` to align with the Azure documentation, this can be changed of course.

The prior PR https://github.com/hashicorp/terraform-provider-azurerm/pull/20619 has mentioned to add this option, but it has not added the option. Instead it has added `ExpressRouteGatewayBypass`.

The option exists for a long time already and is officially documented in the API specs: https://learn.microsoft.com/en-us/azure/templates/microsoft.network/2022-07-01/expressroutegateways?pivots=deployment-language-terraform

I have used the documentation from https://learn.microsoft.com/en-us/azure/templates/microsoft.network/expressroutegateways?pivots=deployment-language-terraform#expressrouteconnectionproperties-2 to supply the necessary documentation.

## PR Checklist

- [x] I have followed the guidelines in our [Contributing Documentation](../blob/main/contributing/README.md).
- [x] I have checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change.
- [x] I have checked if my changes close any open issues. If so please include appropriate [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) below.
- [x] I have updated/added Documentation as required written in a helpful and kind way to assist users that may be unfamiliar with the resource / data source.
- [x] I have used a meaningful PR title to help maintainers and other users understand this change and help prevent duplicate work. 
For example: “`resource_name_here` - description of change e.g. adding property `new_property_name_here`”

## Changes to existing Resource / Data Source

- [x] I have added an explanation of what my changes do and why I'd like you to include them (This may be covered by linking to an issue above, but may benefit from additional explanation).
- [x] I have written new tests for my resource or datasource changes & updated any relevent documentation.
- [ ] I have successfully run tests with my changes locally. If not, please provide details on testing challenges that prevented you running the tests.
- [ ] (For changes that include a **state migration only**). I have manually tested the migration path between relevant versions of the provider.

## Testing 

- [x] My submission includes Test coverage as described in the [Contribution Guide](../blob/main/contributing/topics/guide-new-resource.md) and the tests pass. (if this is not possible for any reason, please include details of why you did or could not add test coverage)

## Change Log

Below please provide what should go into the changelog (if anything) conforming to the [Changelog Format documented here](../blob/main/contributing/topics/maintainer-changelog.md).

* `azurerm_express_route_connection` - support for the `enable_private_link_fast_path` property

This is a (please select all that apply):

- [ ] Bug Fix
- [ ] New Feature (ie adding a service, resource, or data source)
- [x] Enhancement
- [ ] Breaking Change

### CODE REVIEW

File: internal/services/network/express_route_connection_resource.go [hunk 1]
```diff
@@ -71,6 +71,12 @@ func resourceExpressRouteConnection() *pluginsdk.Resource {
 				Optional: true,
 			},
 
+			"enable_private_link_fast_path": {
```
Comment (stephybun) [inline internal/services/network/express_route_connection_resource.go:74]: Please rename this to 
```suggestion
			"private_link_fast_path_enabled": {
```

File: internal/services/network/express_route_connection_resource.go [hunk 2]
```diff
@@ -231,6 +238,10 @@ func resourceExpressRouteConnectionRead(d *pluginsdk.ResourceData, meta interfac
 		d.Set("authorization_key", props.AuthorizationKey)
 		d.Set("enable_internet_security", props.EnableInternetSecurity)
 
+		if props.EnablePrivateLinkFastPath != nil {
+			d.Set("enable_private_link_fast_path", props.EnablePrivateLinkFastPath)
+		}
```
Comment (stephybun) [inline internal/services/network/express_route_connection_resource.go:243]: We could simplify this to 
```suggestion
		d.Set("enable_private_link_fast_path", pointer.From(props.EnablePrivateLinkFastPath))
```

File: website/docs/r/express_route_connection.html.markdown [hunk 3]
```diff
@@ -96,6 +96,8 @@ The following arguments are supported:
 
 * `enable_internet_security` - (Optional) Is Internet security enabled for this Express Route Connection?
 
+* `enable_private_link_fast_path` - (Optional) Bypass the ExpressRoute gateway when accessing private-links. ExpressRoute FastPath (`express_route_gateway_bypass_enabled`) must be enabled.
```
Comment (stephybun) [inline website/docs/r/express_route_connection.html.markdown:99]: What do you mean by 
> ExpressRoute FastPath must be enabled?

File: internal/services/network/express_route_connection_resource_test.go [hunk 4]
```diff
@@ -223,6 +223,7 @@ resource "azurerm_express_route_connection" "test" {
   routing_weight                       = 2
   authorization_key                    = "90f8db47-e25b-4b65-a68b-7743ced2a16b"
   enable_internet_security             = true
+  enable_private_link_fast_path        = true
```
Comment (stephybun) [inline internal/services/network/express_route_connection_resource_test.go:226]: This should also be added to the `complete` test config.

**Review by stephybun — CHANGES_REQUESTED** [review]
Thanks for this @fjaeckel. I left some comments and questions in-line. If you can take a look through those then we'll be happy to give this another review.

File: website/docs/r/express_route_connection.html.markdown (hunk 3 above)
Comment (fjaeckel) [reply website/docs/r/express_route_connection.html.markdown:99]: The other option `express_route_gateway_bypass_enabled` must be set to `true` for the `private_link_fast_path_enabled` to be successful, both options must be set to `true` according to azure docs: https://learn.microsoft.com/en-us/azure/templates/microsoft.network/expressroutegateways?pivots=deployment-language-terraform#expressrouteconnectionproperties-2

File: internal/services/network/express_route_connection_resource_test.go (hunk 4 above)
Comment (fjaeckel) [reply internal/services/network/express_route_connection_resource_test.go]: handled in [fc10726](https://github.com/hashicorp/terraform-provider-azurerm/pull/25596/commits/fc107262eb9afccf873dd7af1395458311eb17e2)

File: internal/services/network/express_route_connection_resource.go (hunk 2 above)
Comment (fjaeckel) [reply internal/services/network/express_route_connection_resource.go]: I've refactored this in https://github.com/hashicorp/terraform-provider-azurerm/pull/25596/commits/3e0b03326152b1f8ed6bfaa6e585ad3467234216

File: internal/services/network/express_route_connection_resource.go (hunk 1 above)
Comment (fjaeckel) [reply internal/services/network/express_route_connection_resource.go]: Renamed in all places.

File: website/docs/r/express_route_connection.html.markdown (hunk 3 above)
Comment (stephybun) [reply website/docs/r/express_route_connection.html.markdown]: In that case can we format the property description for consistency
```
* `enable_private_link_fast_path` - (Optional) Bypass the Express Route gateway when accessing private-links. When enabled `express_route_gateway_bypass_enabled` must be set to `true`. Defaults to `false`.
```

In addition we should probably add some validation within the create for this case?

File: internal/services/network/express_route_connection_resource.go [hunk 5]
```diff
@@ -183,6 +189,7 @@ func resourceExpressRouteConnectionCreate(d *pluginsdk.ResourceData, meta interf
 			RoutingConfiguration:      expandExpressRouteConnectionRouting(d.Get("routing").([]interface{})),
 			RoutingWeight:             utils.Int32(int32(d.Get("routing_weight").(int))),
 			ExpressRouteGatewayBypass: utils.Bool(d.Get("express_route_gateway_bypass_enabled").(bool)),
+			PrivateLinkFastPath:       utils.Bool(d.Get("private_link_fast_path_enabled").(bool)),
```
Comment (stephybun) [inline internal/services/network/express_route_connection_resource.go:192]: We should add some validation here e.g. 

```
	if d.Get("private_link_fast_path_enabled").(bool) && !d.Get("express_route_gateway_bypass_enabled").(bool) {
		return fmt.Errorf("`express_route_gateway_bypass_enabled` must be enabled when `private_link_fast_path_enabled` is set to `true`")
	}
```

Comment (fjaeckel) [reply internal/services/network/express_route_connection_resource.go:197]: I've tried my best putting that at the right position, which should be right before we set the parameters. I've also added it for both the `Create` and `Update` function, so we keep consistency.

File: website/docs/r/express_route_connection.html.markdown (hunk 3 above)
Comment (fjaeckel) [reply website/docs/r/express_route_connection.html.markdown]: Good catch with the validation, that is now being implemented. I've also updated the documentation as you proposed.

**Comment by fjaeckel** [comment]
Tests failed prior, because I accidentally changed the name of the Azure side Parameter, even though it was correct the first time. Fixed in https://github.com/hashicorp/terraform-provider-azurerm/pull/25596/commits/6ce4de4984a4f685128b44fa4292cb16f2fdd00b

**Review by stephybun — APPROVED** [review]
Thanks @fjaeckel LGTM 👍

**Pull request merged**